core/src/main/java/org/apache/calcite/rel/rules/SortRemoveRedundantRule.java

```diff
@@ -96,19 +109,33 @@ protected SortRemoveRedundantRule(final SortRemoveRedundantRule.Config config) {
     }
   }
 
-  private Optional<Double> getSortInputSpecificMaxRowCount(Sort sort) {
-    // If the sort is pure limit, the specific max row count is limit's fetch.
-    if (RelOptUtil.isPureLimit(sort)) {
-      final double limit =
-          sort.fetch instanceof RexLiteral ? RexLiteral.intValue(sort.fetch) : -1D;
-      return Optional.of(limit);
+  private Optional<Integer> getSortInputSpecificMaxRowCount(Sort sort) {
+    if (RelOptUtil.isLimit(sort)) {
+      return getMaxRowCountForLimitAccordingToIsOrder(sort);
     } else if (RelOptUtil.isPureOrder(sort)) {
-      // If the sort is pure order by, the specific max row count is 1.
-      return Optional.of(1D);
+      return Optional.of(1);
     }
     return Optional.empty();
   }
 
+  private Optional<Integer> getMaxRowCountForLimitAccordingToIsOrder(final Sort sort) {
+    final int fetch =
+        sort.fetch instanceof RexLiteral ? RexLiteral.intValue(sort.fetch) : 0;
+
+    // We don't need to deal with fetch is 0.
+    if (fetch == 0) {
+      return Optional.empty();
+    }
+
+    // If sort is 'order by x limit n', the target max row count is 1.
```

Comment:
Why it is 1?

Reply:
Oh, I understand.

You want `inputRowCount <= 1`, but the max count is confusing.

Reply:
I think `threshold` is better than `max row count`.

Reply:
It make sense to me, I will take it.